### ISSUE TITLE
🐛 Diagnostics for redeclared variables

### ISSUE THREAD

**Issue by yassere** [rome/tools#734]
Rome doesn't currently produce diagnostics for:

``` 
const foo;
const foo;
let foo;
let foo;
function foo() {}
function foo() {}
foo;
```

I think we should have a lint rule and/or parser diagnostics for redeclared variables.

**Comment by sebmck** [comment]
Yeah we don't currently do any scope validation, which is bad. I'm not sure where it should be handled. ie. in the parser, lint rules, or during scope construction. There are pros and cons to any of those.

**Comment by yassere** [comment]
Do we want to distinguish between duplicate declarations that throw a syntax error vs duplicate declarations that you can technically do but shouldn't?

**Comment by sebmck** [comment]
We should probably early error for everything the spec says. Meaning implementing the rules in the parser and just emit diagnostics that we could alternatively ignore later. Do you think there would be any value in making the validation generic? ie. would we want to validate a transformed AST to see if it passes the scoping rules?

**Comment by yassere** [comment]
I don't think I've explored the parser closely enough yet to know the best approach. I can take a closer look at this later if nobody else chooses to take this on.

**Comment by jamiebuilds** [comment]
Hi! Thank you for opening this issue

We recently made the decision to [migrate the Rome codebase to Rust](https://rome.tools/blog/2021/09/21/rome-will-be-rewritten-in-rust) and making some significant changes to its architecture. We've also decided not to release the previous version written in TypeScript.

We're going through and closing pull requests and issues that were specific to the TypeScript codebase, including this one. Thank you for taking the time to open it.

Thanks again!
